File: src/purra/agent_tree_execution.py
"""Command and execution boundary for recursive Agent Runs."""

from __future__ import annotations

import asyncio
from dataclasses import dataclass, replace
from typing import Any, Protocol, runtime_checkable
from uuid import uuid4

from purra.agent_tree import (
    AgentNode,
    AgentCapabilityGrant,
    AgentRunAggregation,
    AgentTreeRun,
    AgentTreeRunStatus,
    BeginRootAgentCommand,
    ContinueAgentCommand,
    ContinueAgentReceipt,
    ContextCheckpoint,
    RunTreeRepository,
    SpawnAgentsCommand,
    SpawnAgentsReceipt,
)
from purra.cancellation import OperationCanceled, await_with_cancellation, is_canceled
from purra.agent_tree.delivery import deliver_agent_results
from purra.agent_tree.query import AgentTreeQuery
from purra.agent_tree.receiver import AgentResultReceiver
from purra.agent_tree.lease import bind_agent_run_lease
from purra.interaction import UserInputRequired
from purra.approvals import ApprovalRequired
from purra.errors import ContractViolationError
from purra.normalization import positive_int, required_text
from purra.ports import CancellationSignal


@dataclass(frozen=True, slots=True)
class AgentTreeExecutionResult:
    """Terminal material returned by the one shared Run executor."""

    status: AgentTreeRunStatus
    result: Any = None
    content_ref: str | None = None
    fingerprint: str | None = None
    error_code: str | None = None

    def __post_init__(self) -> None:
        status = AgentTreeRunStatus(self.status)
        if status not in {
            AgentTreeRunStatus.DONE,
            AgentTreeRunStatus.FAILED,
            AgentTreeRunStatus.CANCELED,
        }:
            raise ValueError("Agent tree executor must return a terminal status")
        error = str(self.error_code or "").strip() or None
        if status is AgentTreeRunStatus.DONE:
            if error is not None:
                raise ValueError("completed Agent Run cannot carry an error")
            object.__setattr__(
                self,
                "content_ref",
                required_text(self.content_ref, "Agent Run content reference"),
            )
            object.__setattr__(
                self,
                "fingerprint",
                required_text(self.fingerprint, "Agent Run context fingerprint"),
            )
        elif error is None:
            raise ValueError("unfinished Agent Run requires an error code")
        object.__setattr__(self, "status", status)
        object.__setattr__(self, "error_code", error)


@runtime_checkable
class AgentTreeRunExecutor(Protocol):
    """Execute Root and Child Runs through the same runtime composition."""

    async def execute(
        self,
        run: AgentTreeRun,
        agent: AgentNode,
        checkpoint: ContextCheckpoint | None,
        signal: CancellationSignal | None = None,
    ) -> AgentTreeExecutionResult: ...


@runtime_checkable
class AgentTreeRunCoordinator(Protocol):
    async def execute_and_join(
        self,
        requester_run_id: str,
        run_ids: tuple[str, ...],
        signal: CancellationSignal | None = None,
        *,
        lease_owner_id: str | None = None,
        lease_epoch: int | None = None,
        delivery_signal: CancellationSignal | None = None,
    ) -> AgentRunAggregation: ...


class _AgentTreeJoinCanceled(asyncio.CancelledError):
    code = "child_run_join_canceled"


class _AgentTreeSchedulingCapability:
    """Claim queued Runs, settle them once, and resume their waiting parent."""

    def __init__(
        self,
        *,
        repository: RunTreeRepository,
        executor: AgentTreeRunExecutor,
        owner_id: str | None = None,
        lease_duration_ms: int = 30_000,
        deliver_results=None,
    ) -> None:
        if not isinstance(repository, RunTreeRepository):
            raise TypeError("Agent tree supervisor requires RunTreeRepository")
        if not isinstance(executor, AgentTreeRunExecutor):
            raise TypeError("Agent tree supervisor requires one Run executor")
        self._repository = repository
        self._executor = executor
        self._deliver_results = deliver_results
        self._owner_id = required_text(
            owner_id or f"agent-tree-supervisor-{uuid4().hex}",
            "Agent tree supervisor owner id",
        )
        self._lease_duration_ms = positive_int(
            lease_duration_ms,
            "Agent tree lease duration",
        )

    async def execute_and_join(self, requester_run_id, run_ids, signal=None, *,
                               lease_owner_id=None, lease_epoch=None,
                               delivery_signal=None):
        requester = await self._repository.get_run(requester_run_id)
        if self._deliver_results is not None and requester.root_run_id == requester_run_id:
            return await deliver_agent_results(
                lambda notify: self._execute_and_join(requester_run_id, run_ids, signal,
                    lease_owner_id=lease_owner_id, lease_epoch=lease_epoch, notify=notify),
                lambda results: self._deliver_results(requester_run_id, results, delivery_signal or signal),
            )
        return await self._execute_and_join(
            requester_run_id, run_ids, signal,
            lease_owner_id=lease_owner_id, lease_epoch=lease_epoch,
        )

    async def _execute_and_join(
        self,
        requester_run_id: str,
        run_ids: tuple[str, ...],
        signal: CancellationSignal | None = None,
        *,
        lease_owner_id: str | None = None,
        lease_epoch: int | None = None,
        notify=None,
    ) -> AgentRunAggregation:
        requester_id = required_text(requester_run_id, "requester Run id")
        target_ids = tuple(dict.fromkeys(
            required_text(item, "Child Run id") for item in run_ids
        ))
        if not target_ids:
            return await self._repository.aggregate_runs(requester_id, ())

        requester = await self._repository.get_run(requester_id)
        for identity in target_ids:
            target = await self._repository.get_run(identity)
            for dependency_id in target.dependency_run_ids:
                dependency = await self._repository.get_run(dependency_id)
                if not dependency.terminal and dependency_id not in target_ids:
                    raise ContractViolationError("Join must include unfinished dependencies", code="agent_dependency_join_incomplete")
        if requester.status is AgentTreeRunStatus.WAITING:
            await self._repository.require_run_claim(
                requester_id,
                lease_owner_id=lease_owner_id,
                lease_epoch=lease_epoch,
            )
        else:
            requester = await self._mark_waiting(
                requester_id,
                lease_owner_id=lease_owner_id,
                lease_epoch=lease_epoch,
            )
        root_run_id = requester.root_run_id
        pending = set(target_ids)
        active: dict[asyncio.Task[None], str] = {}
        attempted: set[str] = set()
        joined = False
        try:
            while pending:
                if is_canceled(signal):
                    raise _AgentTreeJoinCanceled(
                        "Child Run join was canceled"
                    )
                aggregate = await self._repository.aggregate_runs(
                    requester_id,
                    target_ids,
                )
                if notify is not None:
                    await notify(aggregate)
                pending = set(aggregate.pending_run_ids)
                if not pending:
                    joined = True
                    return aggregate
                if (await self._repository.get_run(requester_id)).status is AgentTreeRunStatus.RUNNING:
                    await self._mark_waiting(requester_id,
                        lease_owner_id=lease_owner_id, lease_epoch=lease_epoch)
                for candidate in await self._repository.list_runnable(root_run_id):
                    if candidate.run_id not in pending or candidate.run_id in attempted:
                        continue
                    claimed = await self._repository.claim_run(
                        candidate.run_id,
                        owner_id=self._owner_id,
                        lease_duration_ms=self._lease_duration_ms,
                    )
                    if claimed is None:
                        continue
                    async def execute_and_notify(run):
                        await self._execute_claimed(run, signal)
                        if notify is not None:
                            current = await self._repository.aggregate_runs(requester_id, target_ids)
                            await notify(replace(current, results=tuple(item for item in current.results if item["runId"] == run.run_id)))
                    task = asyncio.create_task(execute_and_notify(claimed))
                    active[task] = claimed.run_id
                    attempted.add(claimed.run_id)
                if not active:
                    blocked = set(attempted)
                    changed = True
                    while changed:
                        changed = False
                        for identity in pending - blocked:
                            candidate = await self._repository.get_run(identity)
                            if any(dependency in blocked for dependency in candidate.dependency_run_ids):
                                blocked.add(identity)
                                changed = True
                    if pending <= blocked:
                        joined = True
                        return aggregate
                    # Other joins can occupy the shared root capacity. Their
                    # completion is progress even when this join has no worker.
                    if any(item.status is AgentTreeRunStatus.RUNNING
                           for item in await self._repository.list_descendants(root_run_id)):
                        await asyncio.sleep(0.01)
                        continue
                    raise ContractViolationError(
                        "Child Run scheduler made no progress",
                        code="agent_run_scheduler_stalled",
                    )
                try:
                    done, _ = await await_with_cancellation(
                        asyncio.wait(
                            tuple(active),
                            return_when=asyncio.FIRST_COMPLETED,
                        ),
                        signal,
                    )
                except OperationCanceled:
                    raise _AgentTreeJoinCanceled(
                        "Child Run join was canceled"
                    ) from None
                for task in done:
                    run_id = active.pop(task)
                    await task
                    pending.discard(run_id)
            aggregate = await self._repository.aggregate_runs(
                requester_id,
                target_ids,
            )
            if notify is not None:
                await notify(aggregate)
            joined = True
            return aggregate
        except asyncio.CancelledError as error:
            for run_id in target_ids:
                await self._repository.cancel_subtree(run_id)
            for task in active:
                task.cancel()
            if active:
                await asyncio.gather(*active, return_exceptions=True)
            joined = True
            if isinstance(error, _AgentTreeJoinCanceled):
                raise
            raise _AgentTreeJoinCanceled(
                "Child Run join was canceled"
            ) from None
        finally:
            current = await self._repository.get_run(requester_id)
            if joined and current.status is AgentTreeRunStatus.WAITING:
                try:
                    await self._repository.release_waiting(requester_id,
                        lease_owner_id=lease_owner_id, lease_epoch=lease_epoch)
                except ContractViolationError as error:
                    if error.code not in {"agent_run_state_conflict", "agent_capacity_exceeded"}:
                        raise
                    await self._repository.require_run_claim(requester_id,
                        lease_owner_id=lease_owner_id, lease_epoch=lease_epoch)
                    if (await self._repository.get_run(requester_id)).terminal:
                        raise


    async def _mark_waiting(self, run_id, *, lease_owner_id=None, lease_epoch=None):
        try:
            return await self._repository.mark_waiting(run_id,
                lease_owner_id=lease_owner_id, lease_epoch=lease_epoch)
        except ContractViolationError as error:
            if error.code != "agent_run_state_conflict":
                raise
            await self._repository.require_run_claim(run_id,
                lease_owner_id=lease_owner_id, lease_epoch=lease_epoch)
            current = await self._repository.get_run(run_id)
            if current.status is not AgentTreeRunStatus.WAITING:
                raise
            return current

    async def _execute_claimed(
        self,
        run: AgentTreeRun,
        signal: CancellationSignal | None,
    ) -> None:
        agent = await self._repository.get_agent(run.agent_id)
        checkpoint = (
            await self._repository.get_checkpoint(agent.context_checkpoint_id)
            if agent.context_checkpoint_id is not None
            else None
        )
        expected_context_version = agent.context_version
        try:
            dependencies = [await self._repository.get_run(identity) for identity in run.dependency_run_ids]
            if any(dependency.status is not AgentTreeRunStatus.DONE for dependency in dependencies):
                raise ContractViolationError("A required predecessor did not complete", code="agent_dependency_failed")
            result = await self._execute_with_heartbeat(
                run,
                agent,
                checkpoint,
                signal,
            )
            if not isinstance(result, AgentTreeExecutionResult):
                raise TypeError("Agent tree executor returned an invalid result")
        except asyncio.CancelledError:
            await self._repository.cancel_subtree(run.run_id)
            raise
        except (UserInputRequired, ApprovalRequired):
            await self._repository.suspend_run(run.run_id, lease_owner_id=run.lease_owner_id, lease_epoch=run.lease_epoch)
            return
        except Exception as error:
            await self._repository.fail_run(
                run.run_id,
                str(getattr(error, "code", "") or "agent_execution_failed"),
                lease_owner_id=run.lease_owner_id,
                lease_epoch=run.lease_epoch,
            )
            return
        if result.status is AgentTreeRunStatus.DONE:
            await self._repository.complete_run(
                run.run_id,
                expected_context_version=expected_context_version,
                result=result.result,
                content_ref=result.content_ref or "",
                fingerprint=result.fingerprint or "",
                lease_owner_id=run.lease_owner_id,
                lease_epoch=run.lease_epoch,
            )
        elif result.status is AgentTreeRunStatus.CANCELED:
            await self._repository.cancel_subtree(run.run_id)
        else:
            await self._repository.fail_run(
                run.run_id,
                result.error_code or "agent_run_failed",
                lease_owner_id=run.lease_owner_id,
                lease_epoch=run.lease_epoch,
            )

    async def _execute_with_heartbeat(
        self,
        run: AgentTreeRun,
        agent: AgentNode,
        checkpoint: ContextCheckpoint | None,
        signal: CancellationSignal | None,
    ) -> AgentTreeExecutionResult:
        with bind_agent_run_lease(run.run_id, required_text(run.lease_owner_id, "Agent Run lease owner"), run.lease_epoch):
            execution = asyncio.create_task(self._executor.execute(
                run, agent, checkpoint, signal,
            ))
        heartbeat = asyncio.create_task(self._heartbeat(run))
        try:
            done, _ = await asyncio.wait(
                (execution, heartbeat),
                return_when=asyncio.FIRST_COMPLETED,
            )
            if heartbeat in done:
                error = heartbeat.exception()
                execution.cancel()
                await asyncio.gather(execution, return_exceptions=True)
                if error is not None:
                    raise error
                raise ContractViolationError(
                    "Agent Run heartbeat stopped",
                    code="agent_run_lease_lost",
                )
            return await execution
        finally:
            heartbeat.cancel()
            await asyncio.gather(heartbeat, return_exceptions=True)

    async def _heartbeat(self, run: AgentTreeRun) -> None:
        if run.lease_owner_id is None:
            return
        interval = max(0.001, self._lease_duration_ms / 3_000)
        while True:
            await asyncio.sleep(interval)
            await self._repository.renew_run_lease(
                run.run_id,
                owner_id=run.lease_owner_id,
                lease_epoch=run.lease_epoch,
                lease_duration_ms=self._lease_duration_ms,
            )


class AgentTreeRunSupervisor:
    """Standalone owner for hosts using RunCommandService without AgentCore."""

    def __init__(
        self,
        *,
        repository: RunTreeRepository,
        executor: AgentTreeRunExecutor,
        deliver_results=None,
        owner_id: str | None = None,
        lease_duration_ms: int = 30_000,
    ) -> None:
        self._scheduling = _AgentTreeSchedulingCapability(
            repository=repository,
            executor=executor,
            deliver_results=deliver_results,
            owner_id=owner_id,
            lease_duration_ms=lease_duration_ms,
        )

    async def execute_and_join(
        self,
        requester_run_id: str,
        run_ids: tuple[str, ...],
        signal: CancellationSignal | None = None,
        *,
        lease_owner_id: str | None = None,
        lease_epoch: int | None = None,
        delivery_signal: CancellationSignal | None = None,
    ) -> AgentRunAggregation:
        return await self._scheduling.execute_and_join(
            requester_run_id,
            run_ids,
            signal,
            lease_owner_id=lease_owner_id,
            lease_epoch=lease_epoch,
            delivery_signal=delivery_signal,
        )


class RunCommandService:
    """The only public command path that creates or links Agent Runs."""

    def __init__(
        self,
        repository: RunTreeRepository,
        supervisor: AgentTreeRunCoordinator | None = None,
    ) -> None:
        if not isinstance(repository, RunTreeRepository):
            raise TypeError("Run command service requires RunTreeRepository")
        if supervisor is not None and not isinstance(
            supervisor,
            AgentTreeRunCoordinator,
        ):
            raise TypeError("Run command service supervisor is invalid")
        self._repository = repository
        self._supervisor = supervisor
        self.agents = AgentTreeQuery(repository)
        self.results = AgentResultReceiver(repository, self.join_runs)

    async def begin_root(self, command: BeginRootAgentCommand) -> AgentTreeRun:
        return await self._repository.begin_root(command)

    async def spawn_agents(
        self,
        command: SpawnAgentsCommand,
    ) -> SpawnAgentsReceipt:
        return await self._repository.spawn_agents(command)

    async def compile_child_grant(
        self,
        parent_run_id: str,
        *,
        can_spawn_agents: bool,
        allowed_tools: tuple[str, ...] | None = None,
    ) -> AgentCapabilityGrant:
        if not isinstance(can_spawn_agents, bool):
            raise TypeError("child spawn authority must be boolean")
        parent_run = await self._repository.get_run(parent_run_id)
        parent = await self._repository.get_agent(parent_run.agent_id)
        selected_tools = (
            parent.capability_grant.allowed_tools
            if allowed_tools is None
            else tuple(
                name
                for name in allowed_tools
                if name in parent.capability_grant.allowed_tools
            )
        )
        return replace(
            parent.capability_grant,
            can_spawn_agents=(
                parent.capability_grant.can_spawn_agents and can_spawn_agents
            ),
            allowed_tools=selected_tools,
        )

    async def continue_agent(
        self,
        command: ContinueAgentCommand,
    ) -> ContinueAgentReceipt:
        return await self._repository.continue_agent(command)

    async def join_runs(
        self,
        requester_run_id: str,
        run_ids: tuple[str, ...],
        signal: CancellationSignal | None = None,
        *,
        lease_owner_id: str | None = None,
        lease_epoch: int | None = None,
        delivery_signal: CancellationSignal | None = None,
    ) -> AgentRunAggregation:
        if self._supervisor is None:
            return await self._repository.aggregate_runs(
                requester_run_id,
                run_ids,
            )
        return await self._supervisor.execute_and_join(
            requester_run_id,
            run_ids,
            signal,
            lease_owner_id=lease_owner_id,
            lease_epoch=lease_epoch,
            delivery_signal=delivery_signal,
        )

    async def receive_runs(
        self, requester_run_id: str, run_ids: tuple[str, ...],
        signal: CancellationSignal | None = None, *,
        after_run_ids: tuple[str, ...] = (),
        lease_owner_id: str | None = None, lease_epoch: int | None = None,
    ) -> AgentRunAggregation:
        return await self.results.receive(
            requester_run_id, run_ids, signal, after_run_ids=after_run_ids,
            lease_owner_id=lease_owner_id, lease_epoch=lease_epoch,
        )

    async def cancel_run(self, run_id: str) -> tuple[str, ...]:
        return await self._repository.cancel_subtree(run_id)

    async def close_agent(self, agent_id: str) -> AgentNode:
        return await self._repository.close_agent(agent_id)


__all__ = [
    "AgentTreeExecutionResult",
    "AgentTreeRunExecutor",
    "AgentTreeRunCoordinator",
    "AgentTreeRunSupervisor",
    "RunCommandService",
]

File: typescript/src/agent-tree-execution.ts
import type { JsonValue } from "./model/types.js";
import { AgentError } from "./shared/errors.js";
import { UserInputRequired } from "./interaction.js";
import {
  AgentCapabilityGrant,
  type AgentNode,
  type AgentRunAggregation,
  type AgentTreeRun,
  type BeginRootAgentCommand,
  type ContextCheckpoint,
  type ContinueAgentCommand,
  type ContinueAgentReceipt,
  type RunTreeRepository,
  type SpawnAgentsCommand,
  type SpawnAgentsReceipt,
} from "./agent-tree.js";
import type { DelegationPolicyOptions } from "./delegation/types.js";

export interface AgentTreeOptions {
  readonly repository: RunTreeRepository;
  readonly policy?: DelegationPolicyOptions;
  readonly rootAgentId?: string;
  readonly capabilityGrant?: AgentCapabilityGrant;
  readonly ownerId?: string;
  readonly leaseDurationMs?: number;
}

export interface AgentTreeExecutionResult {
  readonly status: "done" | "failed" | "canceled";
  readonly result?: JsonValue;
  readonly contentRef?: string;
  readonly fingerprint?: string;
  readonly errorCode?: string;
}

export interface AgentTreeRunExecutor {
  execute(
    run: AgentTreeRun,
    agent: AgentNode,
    checkpoint?: ContextCheckpoint,
    signal?: AbortSignal,
  ): Promise<AgentTreeExecutionResult>;
}

export class AgentTreeRunSupervisor {
  readonly #repository: RunTreeRepository;
  readonly #executor: AgentTreeRunExecutor;
  readonly #ownerId: string;
  readonly #leaseDurationMs: number;
  readonly #executionStops = new Map<string, () => void>();

  public constructor(options: {
    readonly repository: RunTreeRepository;
    readonly executor: AgentTreeRunExecutor;
    readonly ownerId?: string;
    readonly leaseDurationMs?: number;
  }) {
    this.#repository = options.repository;
    this.#executor = options.executor;
    this.#ownerId = requiredText(
      options.ownerId ?? `agent-tree-supervisor-${globalThis.crypto.randomUUID()}`,
      "Agent tree supervisor owner id",
    );
    this.#leaseDurationMs = positive(options.leaseDurationMs ?? 30_000, "lease duration");
  }

  public async executeAndJoin(
    requesterRunId: string,
    runIds: readonly string[],
    signal?: AbortSignal,
    claim: { readonly leaseOwnerId?: string; readonly leaseEpoch?: number } = {},
  ): Promise<AgentRunAggregation> {
    const requesterId = requiredText(requesterRunId, "requester Run id");
    const targets = [...new Set(runIds.map((id) => requiredText(id, "Child Run id")))];
    if (targets.length === 0) {
      return this.#repository.aggregateRuns(requesterId, []);
    }
    let requester = await this.#repository.getRun(requesterId);
    if (requester.status === "waiting") {
      await this.#repository.requireRunClaim(requesterId, claim);
    } else {
      requester = await this.#repository.markWaiting(requesterId, claim);
    }
    const pending = new Set(targets);
    const active = new Map<Promise<void>, string>();
    const attempted = new Set<string>();
    let joined = false;
    try {
      while (pending.size > 0) {
        if (signal?.aborted === true) throw joinCanceledError();
        const aggregate = await this.#repository.aggregateRuns(requesterId, targets);
        pending.clear();
        for (const runId of aggregate.pendingRunIds) pending.add(runId);
        if (pending.size === 0) {
          joined = true;
          return aggregate;
        }
        for (const candidate of await this.#repository.listRunnable(requester.rootRunId)) {
          if (!pending.has(candidate.runId) || attempted.has(candidate.runId)) continue;
          const claimed = await this.#repository.claimRun(candidate.runId, {
            ownerId: this.#ownerId,
            leaseDurationMs: this.#leaseDurationMs,
          });
          if (claimed === undefined) continue;
          const task = this.#executeClaimed(claimed, signal);
          active.set(task, claimed.runId);
          attempted.add(claimed.runId);
        }
        if (active.size === 0) {
          if ([...pending].every(id => attempted.has(id))) { joined = true; return aggregate; }
          throw new AgentError(
            "agent_run_scheduler_stalled",
            "Child Run scheduler made no progress",
          );
        }
        const settled = await raceActive(active, signal);
        active.delete(settled.task);
        pending.delete(settled.runId);
      }
      const aggregate = await this.#repository.aggregateRuns(requesterId, targets);
      joined = true;
      return aggregate;
    } catch (error) {
      if (isAbort(error)) {
        for (const runId of targets) {
          await this.#repository.cancelSubtree(runId);
          this.#executionStops.get(runId)?.();
        }
        for (const task of active.keys()) void task.catch(() => undefined);
        joined = true;
      } else {
        await Promise.allSettled(active.keys());
      }
      if (isAbort(error)) throw joinCanceledError(error);
      throw error;
    } finally {
      const current = await this.#repository.getRun(requesterId);
      if (joined && current.status === "waiting") {
        await this.#repository.releaseWaiting(requesterId, claim);
      }
    }
  }

  async #executeClaimed(run: AgentTreeRun, signal?: AbortSignal): Promise<void> {
    const agent = await this.#repository.getAgent(run.agentId);
    const checkpoint = agent.contextCheckpointId === null
      ? undefined
      : await this.#repository.getCheckpoint(agent.contextCheckpointId);
    const expectedContextVersion = agent.contextVersion;
    let result: AgentTreeExecutionResult;
    try {
      result = validateResult(await this.#executeWithHeartbeat(run, agent, checkpoint, signal));
    } catch (error) {
      if (error instanceof UserInputRequired) {
        await this.#repository.suspendRun(run.runId, { leaseOwnerId: run.leaseOwnerId!, leaseEpoch: run.leaseEpoch });
        return;
      }
      if (isAbort(error)) {
        await this.#repository.cancelSubtree(run.runId);
        throw error;
      }
      await this.#repository.failRun(
        run.runId,
        errorCode(error),
        leaseClaim(run),
      );
      return;
    }
    if (result.status === "done") {
      await this.#repository.completeRun(run.runId, {
        expectedContextVersion,
        result: result.result ?? null,
        contentRef: requiredText(result.contentRef, "Agent Run content reference"),
        fingerprint: requiredText(result.fingerprint, "Agent Run context fingerprint"),
        ...leaseClaim(run),
      });
    } else if (result.status === "canceled") {
      await this.#repository.cancelSubtree(run.runId);
    } else {
      await this.#repository.failRun(
        run.runId,
        requiredText(result.errorCode, "Agent Run error code"),
        leaseClaim(run),
      );
    }
  }

  async #executeWithHeartbeat(
    run: AgentTreeRun,
    agent: AgentNode,
    checkpoint?: ContextCheckpoint,
    signal?: AbortSignal,
  ): Promise<AgentTreeExecutionResult> {
    let stopped = false;
    let wake: (() => void) | undefined;
    let stopExecution: (() => void) | undefined;
    const stoppedExecution = new Promise<never>((_resolve, reject) => {
      stopExecution = () => reject(abortError());
    });
    const requestStop = () => stopExecution?.();
    this.#executionStops.set(run.runId, requestStop);
    const sleep = () => new Promise<void>((resolve) => {
      const timer = setTimeout(() => {
        wake = undefined;
        resolve();
      }, Math.max(1, Math.floor(this.#leaseDurationMs / 3)));
      wake = () => {
        clearTimeout(timer);
        wake = undefined;
        resolve();
      };
    });
    const heartbeat = (async () => {
      while (!stopped) {
        await sleep();
        if (stopped) return new Promise<never>(() => undefined);
        await this.#repository.renewRunLease(run.runId, {
          ownerId: requiredText(run.leaseOwnerId, "Agent Run lease owner"),
          leaseEpoch: run.leaseEpoch,
          leaseDurationMs: this.#leaseDurationMs,
        });
      }
      return new Promise<never>(() => undefined);
    })();
    try {
      return await Promise.race([
        this.#executor.execute(run, agent, checkpoint, signal),
        heartbeat,
        stoppedExecution,
      ]);
    } finally {
      stopped = true;
      wake?.();
      if (this.#executionStops.get(run.runId) === requestStop) {
        this.#executionStops.delete(run.runId);
      }
    }
  }
}

export class RunCommandService {
  readonly #repository: RunTreeRepository;
  readonly #supervisor: AgentTreeRunSupervisor | undefined;

  public constructor(
    repository: RunTreeRepository,
    supervisor?: AgentTreeRunSupervisor,
  ) {
    this.#repository = repository;
    this.#supervisor = supervisor;
  }

  public beginRoot(command: BeginRootAgentCommand): Promise<AgentTreeRun> {
    return this.#repository.beginRoot(command);
  }

  public spawnAgents(command: SpawnAgentsCommand): Promise<SpawnAgentsReceipt> {
    return this.#repository.spawnAgents(command);
  }

  public continueAgent(command: ContinueAgentCommand): Promise<ContinueAgentReceipt> {
    return this.#repository.continueAgent(command);
  }

  public async compileChildGrant(
    parentRunId: string,
    options: {
      readonly canSpawnAgents: boolean;
      readonly allowedTools?: readonly string[];
    },
  ): Promise<AgentCapabilityGrant> {
    if (typeof options.canSpawnAgents !== "boolean") {
      throw new TypeError("Child spawn authority must be boolean");
    }
    const run = await this.#repository.getRun(parentRunId);
    const parent = await this.#repository.getAgent(run.agentId);
    const parentGrant = parent.capabilityGrant;
    const allowedTools = options.allowedTools === undefined
      ? parentGrant.allowedTools
      : options.allowedTools.filter((name) => parentGrant.allowedTools.includes(name));
    return new AgentCapabilityGrant({
      canSpawnAgents: parentGrant.canSpawnAgents && options.canSpawnAgents,
      maxDepth: parentGrant.maxDepth,
      maxChildrenPerCall: parentGrant.maxChildrenPerCall,
      maxAgentsPerRoot: parentGrant.maxAgentsPerRoot,
      maxParallelRuns: parentGrant.maxParallelRuns,
      allowedTools,
      allowedModels: parentGrant.allowedModels,
    });
  }

  public joinRuns(
    requesterRunId: string,
    runIds: readonly string[],
    signal?: AbortSignal,
    claim: { readonly leaseOwnerId?: string; readonly leaseEpoch?: number } = {},
  ): Promise<AgentRunAggregation> {
    if (this.#supervisor === undefined) {
      return this.#repository.aggregateRuns(requesterRunId, runIds);
    }
    return this.#supervisor.executeAndJoin(requesterRunId, runIds, signal, claim);
  }

  public cancelRun(runId: string): Promise<readonly string[]> {
    return this.#repository.cancelSubtree(runId);
  }

  public closeAgent(agentId: string): Promise<AgentNode> {
    return this.#repository.closeAgent(agentId);
  }
}

function validateResult(value: AgentTreeExecutionResult): AgentTreeExecutionResult {
  if (value === null || typeof value !== "object") {
    throw new TypeError("Agent tree executor returned an invalid result");
  }
  if (value.status !== "done" && value.status !== "failed" && value.status !== "canceled") {
    throw new TypeError("Agent tree executor must return a terminal status");
  }
  if (value.status === "done") {
    requiredText(value.contentRef, "Agent Run content reference");
    requiredText(value.fingerprint, "Agent Run context fingerprint");
    if (value.errorCode !== undefined) {
      throw new TypeError("Completed Agent Run cannot carry an error");
    }
  } else {
    requiredText(value.errorCode, "Agent Run error code");
  }
  return value;
}

function requiredText(value: unknown, label: string): string {
  const text = typeof value === "string" ? value.trim() : "";
  if (text === "") throw new TypeError(`${label} must be non-empty text`);
  return text;
}

function positive(value: unknown, label: string): number {
  if (!Number.isSafeInteger(value) || Number(value) < 1) {
    throw new TypeError(`${label} must be a positive integer`);
  }
  return Number(value);
}

function leaseClaim(run: AgentTreeRun): {
  readonly leaseOwnerId?: string;
  readonly leaseEpoch?: number;
} {
  return run.leaseOwnerId === null
    ? Object.freeze({})
    : Object.freeze({
        leaseOwnerId: run.leaseOwnerId,
        leaseEpoch: run.leaseEpoch,
      });
}

function abortError(): Error {
  return new DOMException("Agent Run was canceled", "AbortError");
}

function isAbort(error: unknown): boolean {
  return (
    error instanceof DOMException && error.name === "AbortError"
  ) || (
    error instanceof AgentError && error.code === "child_run_join_canceled"
  );
}

function joinCanceledError(cause?: unknown): AgentError {
  return new AgentError(
    "child_run_join_canceled",
    "Child Run join was canceled",
    cause === undefined ? undefined : { cause },
  );
}

function errorCode(error: unknown): string {
  if (error instanceof AgentError) return error.code;
  if (error instanceof Error && error.name !== "") return error.name;
  return "agent_run_failed";
}

async function raceActive(
  active: ReadonlyMap<Promise<void>, string>,
  signal?: AbortSignal,
): Promise<{ readonly task: Promise<void>; readonly runId: string }> {
  const completions = [...active].map(async ([task, runId]) => {
    await task;
    return { task, runId };
  });
  if (signal === undefined) return Promise.race(completions);
  let onAbort: (() => void) | undefined;
  const aborted = new Promise<never>((_resolve, reject) => {
    onAbort = () => reject(joinCanceledError());
    signal.addEventListener("abort", onAbort, { once: true });
    if (signal.aborted) onAbort();
  });
  try {
    return await Promise.race([...completions, aborted]);
  } finally {
    if (onAbort !== undefined) signal.removeEventListener("abort", onAbort);
  }
}
